File: src/types/openai.ts
/**
 * Common OpenAI error shape.
 */
export interface OpenAIError {
    message: string;
    type?: string;
    code?: string;
}

/**
 * Embeddings request.
 */
export interface OpenAIEmbeddingRequest {
    model: string;
    input: string | string[];
}

/**
 * Single embedding object.
 */
export interface OpenAIEmbedding {
    object: 'embedding';
    embedding: number[];
    index: number;
}

/**
 * Embeddings response.
 */
export interface OpenAIEmbeddingResponse {
    object: 'list';
    data: OpenAIEmbedding[];
    model: string;
}

/**
 * Streaming chunk format for chat completions.
 */
export interface OpenAIChatCompletionChunk {
    id: string;
    object: 'chat.completion.chunk';
    created: number;
    model: string;
    choices: Array<{
        index: number;
        delta: {
            role?: 'assistant';
            content?: string;
        };
        finish_reason?: 'stop' | 'length';
    }>;
}

File: rust/rlm/src/types/openai.rs

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenAIError {
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub code: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenAIEmbeddingRequest {
    pub model: String,
    pub input: OpenAIEmbeddingInput,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum OpenAIEmbeddingInput {
    Single(String),
    Multiple(Vec<String>),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenAIEmbedding {
    pub object: String,
    pub embedding: Vec<f32>,
    pub index: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenAIEmbeddingResponse {
    pub object: String,
    pub data: Vec<OpenAIEmbedding>,
    pub model: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenAIChatCompletionChunk {
    pub id: String,
    pub object: String,
    pub created: i64,
    pub model: String,
    pub choices: Vec<OpenAIChatCompletionChunkChoice>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenAIChatCompletionChunkChoice {
    pub index: usize,
    pub delta: OpenAIChatCompletionDelta,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub finish_reason: Option<OpenAIChatFinishReason>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OpenAIChatCompletionDelta {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub role: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum OpenAIChatFinishReason {
    Stop,
    Length,
}

